### PR TITLE
Лабораторна робота №5

### PR DESCRIPTION
**Pull request by MariaShepelieva**
5-е завдання про порівняльний аналіз зроблю згодом, адже ще працюю над правками попередніх лабораторних.

#### 1. Постановка задачі
**- Опис датасета:**  
  **Файл bank--additional-full.csv** містить інформацію про клієнтів банку для маркетингової кампанії.  

**- Цільова змінна:**  
  **y** — приймає значення yes або no.  

**- Основна задача:**  
  Побудувати прогностичну модель, яка на основі характеристик клієнта прогнозує результат (yes або no).

#### 2. Попередня обробка даних
**1. Завантаження даних:**  
   - Через pandas завантажуємо дані та виводимо перші 5 стрічок (df.head()). Пропущені значення позначені як "unknown" і заміняємо їх на NaN.  
   - Перевірено структуру даних (df.info()), відсутні значення (df.isna().sum()).
Отримали:

звіт1
звіт2

**2. Заповнення пропусків:**  
   Для категорій job, marital, education, default, housing, loan пропущені значення заповнені модою.

image
image

**3. Обробка викидів:**  
   Викиди у змінній age були видалені за допомогою IQR-методу.  

image

**4. Кодування категоріальних змінних:**  
   Використано LabelEncoder для змінних:  
   - Категоріальні: job, marital, education, default, housing, loan, contact, month, day_of_week, poutcome.  
   - Цільова змінна y.

image

**5. Розділення даних:**  
   - Фічі (X) та ціль (y) виділені.  
   - Дані поділені на тренувальну та тестову вибірки (train_test_split) у співвідношенні 80/20.  

**6. Масштабування даних:**  
   Застосовано StandardScaler.

---

##### 3.1. Модель AdaBoost

Спочатку будувала базову модель як  DecisionTreeClassifier з max_depth=3 і 
**- Гіперпараметри для оптимізації:**  
  - n_estimators: [10, 50, 100, 200, 300, 500].  
  - learning_rate: [0.01, 0.1, 0.5, 1, 1.5, 2].  

**- Базова модель:**  
  Використано DecisionTreeClassifier з max_depth=1.
  
**- Гіперпараметри для оптимізації:**  
  - n_estimators: [10, 50, 100, 200, 300, 500].  
  - learning_rate: [0.01, 0.1, 0.5, 1, 1.5, 2].  

**- Валідаційні криві:**  

  Побудовано для кожного гіперпараметра:
  - n_estimators:  

image

    *Висновок:* Найкращі значення в діапазоні 200–300.  
  - learning_rate:  

image

    *Висновок:* Найкраще значення 0.5.  

Але графіки нам також показують, що модель демонструє гарний баланс між тренувальною та валідаційною точністю.

**- Фінальна модель:**  
  Побудовано модель з оптимальними гіперпараметрами:  
  - n_estimators=250, learning_rate=0.5. (як і згадувалось вище)

**- Результати:**  
  - Класифікаційний звіт:
  
image

-  (Accuracy):
Значення 0.91 свідчить, що модель правильно класифікує 91% усіх прикладів. Однак точність може бути оманливою, адже в моєму випадку присутній дисбаланс класів.

-  Оцінка F1-метрики:
Загальний F1 Score становить 0.95, що означає гармонійний баланс між  (precision) і  (recall) для всіх класів.

**Для класу no:**
F1-метрика висока (0.95), що підтверджує, що модель добре справляється з виявленням цього класу.
**Для класу yes:**
F1-метрика значно нижча (0.48), що вказує на те, що модель має труднощі з класифікацією цього класу.

-  Precision і Recall:

**Для класу no:**
Precision: 0.93 — більшість передбачень як "no" є правильними.
Recall: 0.98 — модель виявляє майже всі випадки класу "no".

**Для класу yes:**
Precision: 0.67 — лише 67% передбачень як "yes" є правильними.
Recall: 0.38 — модель виявляє лише 50% випадків цього класу, пропускаючи значну кількість.

**Macro avg:**
Середня F1-метрика для обох класів — 0.72, що підтверджує дисбаланс між класами.

**Weighted avg:**
Вага враховує частку кожного класу, тому середня F1-метрика становить 0.90, що ближче до результатів класу "no", адже він переважає.

ROC-AUC: 0.94155...

  - Confusion Matrix.

image

Аналіз: 
**True Negatives (TN): 7032**
Модель правильно класифікувала 7032 випадки як "no".
**False Positives (FP): 222**
Модель помилково класифікувала 222 випадки як "yes", хоча насправді клас був "no".
**False Negatives (FN): 445**
Модель помилково передбачила "no", хоча насправді клас був "yes".
**True Positives (TP): 445**
Модель правильно передбачила 445 випадків як "yes".
---

##### 3.2. Модель GradientBoosting
**- Базова модель:**  
В якості базової моделі використовується неглибоке дерево рішень (max_depth=1).

**-  Підбір гіперпараметрів:**

- Дослідження впливу параметрів n_estimators та learning_rate:

    -   Створення валідаційної кривої для n_estimators.

image

      
    -   Створення валідаційної кривої для learning_rate. 

image

Нотатка: Зі збільшенням кількості дерев розрив між кривими стає більшим, що свідчить про погіршення узагальнюючої здатності моделі (overfitting).

**- Фінальна модель:**  
  Побудовано модель з оптимальними гіперпараметрами:  
  - n_estimators=200, learning_rate=0.1.

**- Результати:**  
  - Класифікаційний звіт.  

image

  - Confusion Matrix:
  
image

Аналіз:
**True Negatives (TN): 7119**
Модель правильно класифікувала 7119 прикладів, коли цільовий клас був "no".
**False Positives (FP): 135**
Модель помилково передбачила "yes", хоча насправді клас був "no".
**False Negatives (FN): 581**
Модель передбачила "no", коли клас насправді був "yes".
**True Positives (TP): 309**
Модель правильно передбачила "yes".

---

#### 4. Аналіз результатів

AdaBoost краще підходить для задачі, якщо ключовим є виявлення класу "yes" (збалансованість Precision і Recall для цього класу). Вона має:

-  Краще значення ROC-AUC.
-  Кращий баланс між Precision і Recall для класу "yes".
-  GradientBoosting підійде для задач, де головним є загальна точність або точність класу "no", оскільки має трохи вищий Precision для "yes", але значно нижчий Recall для цього класу.

#### 6. XGBoost, LightGBM.

-  Класифікаційний звіт

image

Метрики моделей XGBoost та LightGBM

**Для класу 0 (no):**

У обох моделей precision, recall, і F1-score однакові — 0.94, 0.97, і 0.96 відповідно.
Тобто обидві моделі дуже добре класифікують приклади класу 0 (no), який є домінуючим (класовий дисбаланс присутній).

**Для класу 1 (yes):**
-  У XGBoost:
Precision = 0.67 (67% передбачень класу 1 правильні).
Recall = 0.53 (53% реальних прикладів класу 1 було правильно передбачено).
F1-score = 0.59 (збалансована метрика між precision та recall).
-  У LightGBM:
Precision = 0.67, Recall = 0.51, F1-score = 0.58.
Результати для класу 1 схожі, але XGBoost має трохи кращий recall і F1-score.

-  Загальні метрики (Accuracy, Macro avg, Weighted avg):
Accuracy: 92% для обох моделей.
Macro avg:
У XGBoost трохи вищий (0.77 проти 0.74), що свідчить про кращу збалансованість між класами.
Weighted avg: обидві моделі мають однакову вагу (0.92), оскільки домінуючий клас 0 має більший вплив.

Обидві моделі демонструють схожу продуктивність, хоча XGBoost трохи краще працює з меншим класом (клас 1 yes) через вищий recall та F1-score.

- Валідаційні криві:

image
image

У XGBoost ми бачимо сильне перенавчання для тренувальної оцінки. Модель добре запам'ятовує тренувальні дані, але це не дуже добре для узагальнення. Щодо крос-валідаційної оцінки, то ми маємо майже стабільну оцінку, яка трохи зменшується після 100. Збільшення кількості дерев трохи погіршує продуктивність.
Отже: Збільшення кількості дерев швидко призводить до перенавчання, тому необхідно використовувати регуляризацію або обмеження глибини дерев. 

У LightGBM бачимо, що модель краще підганяється під тренувальні дані. Так як відрив від валідаційного значення є значним, то ми модемо зробити висновок про явний ризик перенавчання на тестовій вибірці. Щодо крос-валідаційної кривої, то ми маємо спочатку зростання, а потім незначне спадання після 0.1. Це свідчить про те, що при високих значеннях learning_rate ми отримуємо погіршення узагальнюючої здатності моделі. 
отже: LightGBM  є більш чутливим до learning_rate, і оптимальний вибір цього параметра має вирішальне значення.

Обидві моделі демонструють важливість пошуку оптимальних гіперпараметрів для досягнення найкращої продуктивності.

### CODE REVIEW

**Comment by MariaShepelieva** [comment]
> ## Постановка задачі
> 1. Ознайомитись з описом датасета, визначити цільову змінну (target) та основну задачу прогнозної моделі.
> 2. Виконати попередню обробку даних з метою подальшої побудови моделі.
> 3. Побудувати ансамблеву модель AdaBoost, взявши в якості базової моделі лінійну модель або DecisionTree. Підібрати гіперпараметр >n_estimators, learning_rate. Побудувати валідаційну криву, зробити висновок.

Отримала: 

image

###  Validation Curve with AdaBoost

#### Аналіз розподілу та кривих:
1. Training Score (синя крива):
   - Точність на навчальній вибірці зростає зі збільшенням кількості естіматорів.
   - Крива стабільно зростає, не досягаючи насичення навіть при `500` естіматорах, що може вказувати на можливість перенавчання при значній кількості дерев.

2. Validation Score (помаранчева крива):
   - Точність на тестовій вибірці також зростає зі збільшенням кількості естіматорів, але після `100–200` естіматорів вона стабілізується.
   - Деякі коливання помітні після `200` естіматорів, але їх амплітуда незначна.

#### Оптимальні параметри:
- Найкраще співвідношення між навчанням та узагальненням досягається при кількості естіматорів `100–200`, коли тестова точність максимально близька до навчальної.

#### Точність:
- Максимальна точність на валідаційній вибірці сягає приблизно `0.908–0.910.`
- Надмірне збільшення кількості естіматорів після `200` не дає суттєвого приросту валідаційної точності.

---

> 4. Побудувати ансамблеву модель GradientBoosting, взявши в якості базової моделі неглибоке дерево рішень. Підібрати гіперпараметр n_estimators, learnign_rate. Побудувати валідаційну криву, зробити висновок.

Отримала: 

image


### Validation Curve with GradientBoosting

#### Аналіз розподілу та кривих:
1. Training Score (синя крива):
   - Точність на навчальній вибірці стрімко зростає при невеликій кількості естіматорів (до `100`) і далі сповільнюється, наближаючись до насичення.
   - Крива стає майже горизонтальною після `300` естіматорів, що свідчить про досягнення максимальної ефективності.

2. Validation Score (помаранчева крива):
   - Точність на тестовій вибірці також швидко зростає до `100` естіматорів і стабілізується після цього.
   - Розрив між навчальною і валідаційною точністю менший порівняно з AdaBoost.

#### Оптимальні параметри:
- Кількість естіматорів `100–300` забезпечує баланс між навчанням та узагальненням.

#### Точність:
- Максимальна точність на валідаційній вибірці близько 0.910`0.910,`, що є порівнянним із AdaBoost.
- Валідаційна точність стабільніша, що свідчить про меншу чутливість до переобчислення.

---

### Загальні висновки:
1. AdaBoost:
   - Більш чутливий до збільшення кількості естіматорів.
   - Ризик перенавчання більший, оскільки навчальна точність продовжує зростати навіть при `500 `естіматорах, тоді як валідаційна точність насичується.

2. GradientBoosting:
   - Демонструє стабільніше узагальнення.
   - Менший ризик перенавчання, оскільки крива навчання досягає насичення раніше.

3. Рекомендація:
   - Для обох моделей оптимальна кількість естіматорів становить `100–200`. GradientBoosting є більш стабільним вибором у разі обмеження на складність моделі.

---

> 5. Зробити загальний висновок щодо усіх побудованих моделей у роботах 2-5. Яка модель вийшла найкращою та чому, на вашу думку?

На основі порівняння моделей в роботах` 2-5`, **GradientBoosting** виявляється кращою моделлю. Модель має меншу схильність до перенавчання завдяки більш поступовому додаванню нових дерев, що дозволяє ефективно виправляти помилки попередніх. Її стабільність на валідаційних даних свідчить про кращу узагальнюючу здатність, що робить її оптимальним вибором для даної задачі. Але її мінус в тому, що ця модель повільна.

В загальному можемо сказати: 

XGBoost: найкраща точність, підходить для складних даних.
LightGBM: оптимально для швидкого навчання.
GradientBoosting: стабільна, але повільна.
AdaBoost: простий, але менш точний.

---

> 6. (творче завдання) Побудувати модель, використовуючи XGBoost, LightGBM. Підібрати гіперпараметри, побудувати валідаційні криві. Зробити аналіз важливості ознак, побудувати графіки. Зробити висновки.

Отримано результати:


```
Найкращі параметри для XGBoost: {'learning_rate': 0.05, 'max_depth': 5, 'n_estimators': 150}
Точність на валідаційних даних (XGBoost): 0.9167223065250379
[LightGBM] [Warning] min_data_in_leaf is set=50, min_child_samples=20 will be ignored. Current value: min_data_in_leaf=50
[LightGBM] [Warning] min_data_in_leaf is set=50, min_child_samples=20 will be ignored. Current value: min_data_in_leaf=50
[LightGBM] [Info] Number of positive: 3705, number of negative: 29245
[LightGBM] [Info] Total Bins 751
[LightGBM] [Info] Number of data points in the train set: 32950, number of used features: 51
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.112443 -> initscore=-2.066025
[LightGBM] [Info] Start training from score -2.066025
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
Найкращі параметри для LightGBM: {'learning_rate': 0.05, 'max_depth': 5, 'n_estimators': 150}
Точність на валідаційних даних (LightGBM): 0.9166616084977239

```

image
image

**Аналіз валідаційної кривої з LightGBM:**

- Перенавчання:

  - Зі збільшенням кількості дерев точність на тренувальних даних постійно зростає, що свідчить про те, що модель починає запам'ятовувати шум у тренувальних даних, а не вивчати загальні закономірності.
  - Точність на валідаційних даних спочатку зростає, досягає максимуму, а потім починає стабілізуватися або навіть злегка знижуватися. Це класична ознака перенавчання.

- Оптимальна кількість дерев:

  - Оптимальна кількість дерев - це точка, в якій точність на валідаційних даних максимальна. На цьому графіку це приблизно 100-150 дерев.
  - Після цієї точки додавання нових дерев не покращує узагальнюючу здатність моделі і може навіть погіршити її.

- Загальна точність:

  - Максимальна точність на валідаційних даних досить висока, що свідчить про те, що модель загалом добре підходить для вирішення поставленого завдання.


**Аналіз валідаційної кривої з XGBoost:**

- Перенавчання:

  - Зі збільшенням кількості дерев точність на тренувальних даних постійно зростає, що свідчить про те, що модель починає запам'ятовувати шум у тренувальних даних, а не вивчати загальні закономірності. 
  - Точність на валідаційних даних спочатку зростає, досягає максимуму, а потім починає стабілізуватися або навіть злегка знижуватися. Це класична ознака перенавчання.

- Оптимальна кількість дерев:

  - Оптимальна кількість дерев - це точка, в якій точність на валідаційних даних максимальна. На цьому графіку це приблизно 100-150 дерев.
  - Після цієї точки додавання нових дерев не покращує узагальнюючу здатність моделі і може навіть погіршити її.

Загальна точність:

  - Максимальна точність на валідаційних даних досить висока (91.67%), що свідчить про те, що модель загалом добре підходить для вирішення поставленого завдання.

Висновок:

На основі представленого графіка можна зробити висновок, що модель XGBoost здатна досягти високої точності на даних, але при цьому схильна до перенавчання. 

Загальний висновок:

Для отримання оптимальних результатів необхідно ретельно підібрати гіперпараметри моделі і обрізати її, щоб уникнути перенавчання.


Топ-10 важливих ознак для   LightGBM і XGBoost відповідно: 

image

image
